core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java

```diff
@@ -134,31 +137,83 @@ public static boolean isMergeJoinSupported(JoinRelType joinType) {
       final RelTraitSet required) {
     // Required collation keys can be subset or superset of merge join keys.
     RelCollation collation = required.getCollation();
+    int leftInputFieldCount = left.getRowType().getFieldCount();
+
     List<Integer> reqKeys = RelCollations.ordinals(collation);
-    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
+    List<Integer> leftKeys = joinInfo.leftKeys.toIntegerList();
+    List<Integer> rightKeys =
+        joinInfo.rightKeys.incr(leftInputFieldCount).toIntegerList();
+    List<Integer> rightKeysNotShifted = joinInfo.rightKeys.toIntegerList();
 
+    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
     ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
     ImmutableBitSet rightKeySet = ImmutableBitSet.of(joinInfo.rightKeys)
-        .shift(left.getRowType().getFieldCount());
+        .shift(leftInputFieldCount);
 
-    // Only consider exact key match for now
     if (reqKeySet.equals(leftKeySet)) {
+      // if sort keys equal to left join keys, we can pass through all collations directly.
       Mappings.TargetMapping mapping = buildMapping(true);
       RelCollation rightCollation = collation.apply(mapping);
       return Pair.of(
           required, ImmutableList.of(required,
           required.replace(rightCollation)));
+    } else if (containsOrderless(collation, leftKeys)) {
```

Comment:
How can it be true if sort keys are subset of left join keys?

Reply:
yeah I guess I have made this wrong. 

The current MergeJoin implementation construct key comparator just by sort key. If keep code here unchanged, I think I need to update at least this line: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java#L310

Basically if there are collations defined (on join keys), the key comparator (and key selector) will need to be constructed on collations's ordering.

Reply:
You might not notice, I have changed the code of `containsOrderless` and added this
```
    if (colKeys.size() < distinctKeys.size()) {
      return containsOrderlessSubset(collation.getKeys().toIntegerList(), distinctKeys);
    } else {
```
So firstly, collations contains fields less than join keys is allowed. Then, it is only legal if collations follows the ordering of sort keys (and later at runtime, key selector and key comparator are constructed by sort key ordering).

Reply:
This breaks the definition of the method name. `containsOrderless` means the former contains the latter. We can create another method `containsOrderless(List<Integer>, List<Integer>)` if necessary.

Reply:
I need to update `containsOrderlessSubset` to make it consider list ordering.

Reply:
Add `containsOrderless(List<Integer>, List<Integer>)` now to obey the naming convention.